arrabiata/src/interpreter.rs

```diff
@@ -0,0 +1,962 @@
+//! - for each row `i`, we have `x_i = x_{i - 1} << 15 - \sum_{j=0}^{14} 2^j b_{i * 15 + j}`
+//! - `b_i` is the i-th bit of the input value
+//!
+//! ## Handle the combinaison of constraints
+//!
+//! The prover will have to combine the constraints to generate the
+//! full circuit at the end. The constraints will be combined using a
+//! challenge (often called α) that will be generated in the verifier circuit by
+//! simulating the Fiat-Shamir transformation.
+//! The challenges will then be accumulated over time using the random coin used
+//! by the folding argument.
+//! The verifier circuit must be carefully implemented to ensure that all the
+//! messages that the prover would have sent before coining the random combiner
+//! for the constraints has been absorbed properly in the verifier circuit.
+//!
+//! Using this technique requires us a folding scheme that handles degree
+//! `5 + 1` constraints, as the challenge will be considered as a variable.
+//! The reader can refer to the folding library available in this monorepo for
+//! more contexts.
+//!
+//! ## Permutation argument
+//!
+//! Communication between rows must be done using a permutation argument. The
+//! argument we use will be a generalisation of the one used in the [PlonK
+//! paper](https://eprint.iacr.org/2019/953).
+//!
+//! The construction of the permutations will be done using the methods prefixed
+//! `save` and `load`. The index of the current row and the index of the
+//! time the value has been written will be used to generate the permutation on
+//! the fly.
+//!
+//! The permutation argument described in the PlonK paper is a kind of "inverse
+//! lookup" protocol, like Plookup. The polynomials are defined as follows:
+//!
+//! ```text
+//!            Can be seen as T[f(X)] = Χ
+//!          --------
+//!          |      |
+//! f'(X) = f(X) + β X + γ
+//!                      |--- Can be seen as the evaluation point.
+//!                         |
+//!                         |
+//! g'(X) = g(X) + β σ(X) + γ
+//!          |      |
+//!          --------
+//!          Can be seen as T[g(X)] = σ(X)
+//! ```
+//!
+//! And from this, we build an accumulator, like for Plookup.
+//! The accumulator requires to coin two challenges, β and γ, and it must be
+//! done after the commitments to the columns have been absorbed.
+//! The verifier at the next step will verify that the challenges have been
+//! correctly computed.
+//! In the implementation, the accumulator will be computed after the challenges
+//! and the commitments. Note that the accumulator must also be aggregated, and
+//! the aggregation must be performed by the verifier at the next step.
+//!
+//! The methods `save` and `load` will accept as arguments only a column that is
+//! included in the permutation argument. For instance, `save_poseidon_state`
+//! will only accept columns with index 3, 4 and 5, where the
+//! `load_poseidon_state` will only accepts columns with index 0, 1 and 2.
```

Comment:
why are the save and load in different columns ?
if a perm argument  is applied to a set of columns, one can load and save from/to any columns of the set